File: TCPIP/src/22/traceroute_implementation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/udp.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <errno.h>

#define MAX_HOPS 30
#define PACKET_SIZE 64
#define TIMEOUT_SEC 2
#define DEST_PORT 33434  // Standard traceroute port

// Calculate checksum for ICMP packet
unsigned short checksum(void *b, int len) {
    unsigned short *buf = b;
    unsigned int sum = 0;
    unsigned short result;

    for (sum = 0; len > 1; len -= 2)
        sum += *buf++;
    
    if (len == 1)
        sum += *(unsigned char *)buf;
    
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    result = ~sum;
    
    return result;
}

// Send UDP probe packet with specific TTL
int send_probe(int sockfd, struct sockaddr_in *addr, int ttl, int seq) {
    char packet[PACKET_SIZE];
    
    // Set TTL for this packet
    if (setsockopt(sockfd, IPPROTO_IP, IP_TTL, &ttl, sizeof(ttl)) < 0) {
        perror("setsockopt TTL failed");
        return -1;
    }
    
    memset(packet, 0, PACKET_SIZE);
    
    // Fill with sequence number for identification
    *(int *)packet = seq;
    
    struct sockaddr_in dest = *addr;
    dest.sin_port = htons(DEST_PORT + seq);
    
    if (sendto(sockfd, packet, PACKET_SIZE, 0, 
               (struct sockaddr *)&dest, sizeof(dest)) < 0) {
        perror("sendto failed");
        return -1;
    }
    
    return 0;
}

// Receive ICMP response (Time Exceeded or Dest Unreachable)
int receive_response(int icmp_sockfd, struct sockaddr_in *from_addr, 
                     double *rtt, int expected_seq) {
    char buffer[1024];
    struct timeval tv_start, tv_end;
    socklen_t addr_len = sizeof(*from_addr);
    
    gettimeofday(&tv_start, NULL);
    
    int n = recvfrom(icmp_sockfd, buffer, sizeof(buffer), 0,
                     (struct sockaddr *)from_addr, &addr_len);
    
    gettimeofday(&tv_end, NULL);
    
    if (n < 0) {
        if (errno == EAGAIN || errno == EWOULDBLOCK) {
            return -1;  // Timeout
        }
        perror("recvfrom failed");
        return -1;
    }
    
    // Calculate RTT
    *rtt = (tv_end.tv_sec - tv_start.tv_sec) * 1000.0 +
           (tv_end.tv_usec - tv_start.tv_usec) / 1000.0;
    
    // Parse IP header
    struct ip *ip_hdr = (struct ip *)buffer;
    int ip_hdr_len = ip_hdr->ip_hl << 2;
    
    // Parse ICMP header
    struct icmp *icmp_hdr = (struct icmp *)(buffer + ip_hdr_len);
    
    // Check ICMP type
    if (icmp_hdr->icmp_type == ICMP_TIMXCEED) {
        return 0;  // Time exceeded - intermediate router
    } else if (icmp_hdr->icmp_type == ICMP_UNREACH) {
        return 1;  // Destination unreachable - final destination
    }
    
    return -1;  // Other ICMP type
}

// Resolve IP address to hostname
char* reverse_dns(struct in_addr addr) {
    static char hostname[256];
    struct hostent *host;
    
    host = gethostbyaddr(&addr, sizeof(addr), AF_INET);
    
    if (host && host->h_name) {
        snprintf(hostname, sizeof(hostname), "%s (%s)", 
                 host->h_name, inet_ntoa(addr));
    } else {
        snprintf(hostname, sizeof(hostname), "%s", inet_ntoa(addr));
    }
    
    return hostname;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <hostname>\n", argv[0]);
        return 1;
    }
    
    // Resolve hostname
    struct hostent *host = gethostbyname(argv[1]);
    if (!host) {
        fprintf(stderr, "Unknown host: %s\n", argv[1]);
        return 1;
    }
    
    struct sockaddr_in dest_addr;
    memset(&dest_addr, 0, sizeof(dest_addr));
    dest_addr.sin_family = AF_INET;
    dest_addr.sin_addr = *(struct in_addr *)host->h_addr;
    
    printf("traceroute to %s (%s), %d hops max, %d byte packets\n",
           argv[1], inet_ntoa(dest_addr.sin_addr), MAX_HOPS, PACKET_SIZE);
    
    // Create UDP socket for sending probes
    int udp_sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (udp_sockfd < 0) {
        perror("UDP socket creation failed");
        return 1;
    }
    
    // Create raw ICMP socket for receiving responses (requires root)
    int icmp_sockfd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
    if (icmp_sockfd < 0) {
        perror("ICMP socket creation failed (need root privileges)");
        close(udp_sockfd);
        return 1;
    }
    
    // Set timeout on ICMP socket
    struct timeval timeout;
    timeout.tv_sec = TIMEOUT_SEC;
    timeout.tv_usec = 0;
    setsockopt(icmp_sockfd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
    
    // Traceroute loop
    int reached_dest = 0;
    
    for (int ttl = 1; ttl <= MAX_HOPS && !reached_dest; ttl++) {
        printf("%2d  ", ttl);
        fflush(stdout);
        
        int got_response = 0;
        struct sockaddr_in from_addr;
        double rtt;
        
        // Send 3 probes per hop
        for (int probe = 0; probe < 3; probe++) {
            int seq = ttl * 1000 + probe;
            
            if (send_probe(udp_sockfd, &dest_addr, ttl, seq) < 0) {
                printf("* ");
                continue;
            }
            
            int result = receive_response(icmp_sockfd, &from_addr, &rtt, seq);
            
            if (result >= 0) {
                if (!got_response) {
                    // Print hostname/IP for first response
                    char *hostname = reverse_dns(from_addr.sin_addr);
                    printf("%s  ", hostname);
                    got_response = 1;
                }
                
                printf("%.3f ms  ", rtt);
                
                // Check if we reached destination
                if (result == 1 || 
                    from_addr.sin_addr.s_addr == dest_addr.sin_addr.s_addr) {
                    reached_dest = 1;
                }
            } else {
                printf("* ");
            }
            
            fflush(stdout);
        }
        
        printf("\n");
    }
    
    close(udp_sockfd);
    close(icmp_sockfd);
    
    return 0;
}

File: TCPIP/src/22/traceroute_implementation.rs
use std::net::{IpAddr, Ipv4Addr, SocketAddr, ToSocketAddrs};
use std::time::{Duration, Instant};
use std::io;

const MAX_HOPS: u8 = 30;
const PACKET_SIZE: usize = 64;
const TIMEOUT: Duration = Duration::from_secs(2);
const DEST_PORT_BASE: u16 = 33434;

// ICMP header types
const ICMP_ECHOREPLY: u8 = 0;
const ICMP_DEST_UNREACH: u8 = 3;
const ICMP_TIME_EXCEEDED: u8 = 11;

fn send_probe(
    socket: &socket2::Socket,
    dest: &socket2::SockAddr,
    ttl: u8,
    seq: u32,
) -> io::Result<Instant> {
    // Set TTL for this probe
    socket.set_ttl(ttl as u32)?;
    
    // Create probe packet with sequence number
    let mut packet = vec![0u8; PACKET_SIZE];
    packet[0..4].copy_from_slice(&seq.to_be_bytes());
    
    let send_time = Instant::now();
    socket.send_to(&packet, dest)?;
    
    Ok(send_time)
}

fn receive_icmp_response(
    socket: &socket2::Socket,
    send_time: Instant,
) -> io::Result<(Ipv4Addr, Duration, bool)> {
    let mut buffer = [0u8; 1024];
    
    let (len, addr) = socket.recv_from(&mut buffer)?;
    let recv_time = Instant::now();
    let rtt = recv_time.duration_since(send_time);
    
    if len < 28 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "Packet too short",
        ));
    }
    
    // Parse IP header
    let ip_header_len = ((buffer[0] & 0x0F) * 4) as usize;
    
    if len < ip_header_len + 8 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "ICMP packet too short",
        ));
    }
    
    // Extract source IP from IP header (bytes 12-15)
    let src_ip = Ipv4Addr::new(
        buffer[12],
        buffer[13],
        buffer[14],
        buffer[15],
    );
    
    // Parse ICMP header
    let icmp_type = buffer[ip_header_len];
    
    // Determine if we reached the destination
    let is_destination = match icmp_type {
        ICMP_TIME_EXCEEDED => false,      // Intermediate hop
        ICMP_DEST_UNREACH => true,        // Destination reached
        ICMP_ECHOREPLY => true,           // Destination reached (if using ICMP)
        _ => false,
    };
    
    Ok((src_ip, rtt, is_destination))
}

fn resolve_hostname(ip: Ipv4Addr) -> String {
    // Attempt reverse DNS lookup
    match dns_lookup::lookup_addr(&IpAddr::V4(ip)) {
        Ok(hostname) => format!("{} ({})", hostname, ip),
        Err(_) => format!("{}", ip),
    }
}

fn traceroute(dest_ip: Ipv4Addr, dest_name: &str) -> io::Result<()> {
    println!(
        "traceroute to {} ({}), {} hops max, {} byte packets",
        dest_name, dest_ip, MAX_HOPS, PACKET_SIZE
    );
    
    // Create UDP socket for sending probes
    let udp_socket = socket2::Socket::new(
        socket2::Domain::IPV4,
        socket2::Type::DGRAM,
        Some(socket2::Protocol::UDP),
    )?;
    
    // Create raw ICMP socket for receiving responses
    let icmp_socket = socket2::Socket::new(
        socket2::Domain::IPV4,
        socket2::Type::RAW,
        Some(socket2::Protocol::ICMPV4),
    )?;
    
    icmp_socket.set_read_timeout(Some(TIMEOUT))?;
    
    let mut reached_dest = false;
    
    for ttl in 1..=MAX_HOPS {
        if reached_dest {
            break;
        }
        
        print!("{:2}  ", ttl);
        
        let mut hop_responded = false;
        let mut hop_ip: Option<Ipv4Addr> = None;
        
        // Send 3 probes per hop
        for probe in 0..3 {
            let seq = (ttl as u32) * 1000 + probe;
            let port = DEST_PORT_BASE + seq as u16;
            
            let dest_sock_addr = socket2::SockAddr::from(
                SocketAddr::new(IpAddr::V4(dest_ip), port)
            );
            
            match send_probe(&udp_socket, &dest_sock_addr, ttl, seq) {
                Ok(send_time) => {
                    match receive_icmp_response(&icmp_socket, send_time) {
                        Ok((from_ip, rtt, is_dest)) => {
                            // Print hostname on first response from this hop
                            if !hop_responded {
                                let hostname = resolve_hostname(from_ip);
                                print!("{}  ", hostname);
                                hop_responded = true;
                                hop_ip = Some(from_ip);
                            }
                            
                            print!("{:.3} ms  ", rtt.as_secs_f64() * 1000.0);
                            
                            if is_dest || from_ip == dest_ip {
                                reached_dest = true;
                            }
                        }
                        Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
                            print!("* ");
                        }
                        Err(_) => {
                            print!("* ");
                        }
                    }
                }
                Err(_) => {
                    print!("* ");
                }
            }
        }
        
        println!();
    }
    
    Ok(())
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() != 2 {
        eprintln!("Usage: {} <hostname>", args[0]);
        std::process::exit(1);
    }
    
    let host = &args[1];
    
    // Resolve hostname to IP address
    let addr = format!("{}:0", host)
        .to_socket_addrs()?
        .find(|addr| addr.is_ipv4())
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "Host not found"))?;
    
    let dest_ip = match addr.ip() {
        IpAddr::V4(ip) => ip,
        _ => {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "IPv6 not supported",
            ))
        }
    };
    
    traceroute(dest_ip, host)?;
    
    Ok(())
}

// Add to Cargo.toml:
// [dependencies]
// socket2 = "0.5"
// dns-lookup = "2.0"